plasmapy/atomic/classes.py

```diff
@@ -154,21 +109,54 @@ def __init__(self,
                  argument: Union[str, int],
                  mass_numb: int = None,
                  Z: int = None):
-        r"""Initializes a Particle object."""
-
-        # If a private variable is not defined later, then the value of None
-        # lets us know that the corresponding attribute is not available.
-        for variable in _private_variables:
-            if variable not in self.__dict__:
-                exec(f'self.{variable} = None')
+        r"""Initializes a Particle object by setting all necessary private
+        attributes."""
+
+        if not isinstance(argument, (int, str)):
+            raise TypeError(
+                "The first positional argument when creating a Particle "
+                "object must be either an integer or string.")
+
+        # If the data is missing, then the private attribute should still
+        # exist but just be set to None.  This initialization had previously
+        # been done in a loop using exec on a string, but this does not play
+        # well with static type checkers such as PyCharm.
```

Comment:
Good choice going back on that, I think. It was probably too fancy for its own good, though kudos on the idea.